src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -256,6 +258,7 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
                 diagnosticsBag?.AddRange(diagnostics);
             }
             state = state.With(generatorStates: stateBuilder.ToImmutableAndFree());
+            state.ModelProvider.ClearCache(compilation);
```

Comment:
❔ Is this still going to get called if generation is cancelled?

Reply:
Hmm, I think there are some code paths where it does not. Given that we're only ever caching against a single compilation, I think I'll just lift up the `PerCompilationProvider` , skip the compilation -> provider cache, and create a new one per-generation.